src/pool/spawn.rs

```diff
@@ -1,56 +1,276 @@
+///
+/// Note that thread pool can be shutdown and dropped even not all remotes are
+/// dropped.
 pub struct Remote<T> {
-    injector: Arc<TaskInjector<T>>,
+    core: Arc<QueueCore<T>>,
 }
 
-impl<T> Remote<T> {
-    pub(crate) fn new(injector: Arc<TaskInjector<T>>) -> Remote<T> {
-        Remote { injector }
+impl<T: TaskCell + Send> Remote<T> {
+    pub(crate) fn new(core: Arc<QueueCore<T>>) -> Remote<T> {
+        Remote { core }
     }
-}
 
-impl<T: TaskCell> Remote<T> {
-    /// Spawns the tasks into thread pool.
-    ///
-    /// If the thread pool is shutdown, it becomes no-op.
-    pub fn spawn(&self, _t: impl Into<T>) {
-        unimplemented!()
+    /// Submits a task to the thread pool.
+    pub fn spawn(&self, task: impl Into<T>) {
+        self.core.push(0, task.into());
+    }
+
+    pub(crate) fn stop(&self) {
+        self.core.mark_shutdown(0);
     }
 }
 
 impl<T> Clone for Remote<T> {
     fn clone(&self) -> Remote<T> {
         Remote {
-            injector: self.injector.clone(),
+            core: self.core.clone(),
+        }
+    }
+}
+
+/// Note that implements of Runner assumes `Remote` is `Sync` and `Send`.
+/// So we need to use assert trait to ensure the constraint at compile time
+/// to avoid future breaks.
+trait AssertSync: Sync {}
+impl<T: Send> AssertSync for Remote<T> {}
+trait AssertSend: Send {}
+impl<T: Send> AssertSend for Remote<T> {}
+
+/// Spawns tasks to the associated thread pool.
+///
+/// It's different from `Remote` because it submits tasks to the local queue
+/// instead of global queue, so new tasks can take advantage of cache
+/// coherence.
+pub struct Local<T> {
+    id: usize,
+    local_queue: LocalQueue<T>,
+    core: Arc<QueueCore<T>>,
+}
+
+impl<T: TaskCell + Send> Local<T> {
+    pub(crate) fn new(id: usize, local_queue: LocalQueue<T>, core: Arc<QueueCore<T>>) -> Local<T> {
+        Local {
+            id,
+            local_queue,
+            core,
+        }
+    }
+
+    /// Spawns a task to the local queue.
+    pub fn spawn(&mut self, task: T) {
+        self.local_queue.push(task);
+    }
+
+    /// Spawns a task to the remote queue.
+    pub fn spawn_remote(&self, task: T) {
+        self.core.push(self.id, task);
+    }
+
+    /// Gets a remote so that tasks can be spawned from other threads.
+    pub fn remote(&self) -> Remote<T> {
+        Remote {
+            core: self.core.clone(),
+        }
+    }
+
+    pub(crate) fn core(&self) -> &Arc<QueueCore<T>> {
+        &self.core
+    }
+
+    pub(crate) fn pop(&mut self) -> Option<Pop<T>> {
+        self.local_queue.pop()
+    }
+
+    /// Pops a task from the queue.
+    ///
+    /// If there is no tasks at the moment, it will go to sleep until woken
+    /// up by other threads.
+    pub(crate) fn pop_or_sleep(&mut self) -> Option<Pop<T>> {
+        let address = &*self.core as *const QueueCore<T> as usize;
+        let mut task = None;
+        let mut timeout = Some(Instant::now() + self.core.config.max_idle_time);
```

Comment:
I can't understand what the idle time means.
Now it seems to be the maximum time the thread waits for the first time and try to pop again. And at the second time, it waits forever until unparked.